File: src/main/java/travelator/money/ExchangeRates.java
package travelator.money;

import java.math.BigDecimal;
import java.util.Currency;

public interface ExchangeRates {
    BigDecimal rate(Currency fromCurrency, Currency toCurrency);

    default CurrencyConversion convert(Money fromMoney, Currency toCurrency) {
        BigDecimal rate = rate(fromMoney.getCurrency(), toCurrency);
        BigDecimal toAmount = fromMoney.getAmount().multiply(rate);
        Money toMoney = Money.of(toAmount, toCurrency);
        
        return new CurrencyConversion(fromMoney, toMoney);
    }
}

File: src/main/java/travelator/money/ExchangeRates.kt
package travelator.money

import java.math.BigDecimal
import java.util.*

interface ExchangeRates {
    fun rate(fromCurrency: Currency, toCurrency: Currency): BigDecimal

    @JvmDefault
    fun convert(fromMoney: Money, toCurrency: Currency): CurrencyConversion {
        val rate = rate(fromMoney.currency, toCurrency)
        val toAmount = fromMoney.amount * rate
        val toMoney = Money(toAmount, toCurrency)

        return CurrencyConversion(fromMoney, toMoney)
    }
}